rainier-core/src/main/scala/com/stripe/rainier/core/Categorical.scala

```diff
@@ -9,21 +10,33 @@ import com.stripe.rainier.compute._
   */
 final case class Categorical[T](pmf: Map[T, Real]) extends Distribution[T] {
   self =>
+
   def map[U](fn: T => U): Categorical[U] =
-    flatMap { t =>
-      Categorical(Map(fn(t) -> Real.one))
-    }
+    Categorical(
+      pmf.foldLeft(Map.empty[U, Real]) {
+        case (acc, (t, p)) =>
+          updateMap(acc, fn(t), p)(Real.zero)(_ + _)
+      }
+    )
 
   def flatMap[U](fn: T => Categorical[U]): Categorical[U] =
     Categorical(
-      pmf.toList
-        .flatMap {
-          case (t, p) =>
-            fn(t).pmf.toList.map { case (u, p2) => (u, p * p2) }
-        }
-        .groupBy(_._1)
-        .map { case (u, ups) => u -> Real.sum(ups.map(_._2)) }
-        .toMap)
+      (for {
```

Comment:
This implementation is more efficient, as we take fewer passes over each of the `pmf` instances.